### PR TITLE
Adding support for failure metrics

### PR DESCRIPTION
**Pull request by lastmansleeping**
- Support `is_secondary_label` in the feature_config to specify secondary labels to be used to compute failure_metrics
- Computing failure metrics
- Add secondary label data to test/dummy data
- Updating tests
- Add tests for failure metrics

### CODE REVIEW

**Comment by lastmansleeping** [comment]
To test, run:
```
python ml4ir/applications/ranking/pipeline.py \
--data_dir ml4ir/applications/ranking/tests/data/tfrecord \
--feature_config ml4ir/applications/ranking/tests/data/config/feature_config.yaml \
--run_id test \
--data_format tfrecord \
--execution_mode train_inference_evaluate
```

`feat_3` in the feature_config is specified as a secondary_label, which will be used to compute the ranking failure metric. You should see the metrics logged as part of the model evaluation after training.

**Comment by lastmansleeping** [comment]
Expected metrics (the actual numbers might be different):
```
Overall Metrics: 
query_count                                    1408.000000
old_ACR                                           1.650568
new_ACR                                           1.944602
old_MRR                                           0.785881
new_MRR                                           0.704753
old_failure_count_feat_3                          0.063920
new_failure_count_feat_3                          0.084517
old_failure_partial_count_feat_3                  0.112216
new_failure_partial_count_feat_3                  0.161932
old_failure_score_feat_3                          0.166903
new_failure_score_feat_3                          0.250710
old_failure_score_normalized_feat_3               0.088986
new_failure_score_normalized_feat_3               0.125103
perc_improv_ACR                                 -17.814114
perc_improv_MRR                                 -10.323127
perc_improv_failure_count_feat_3                 -0.322222
perc_improv_failure_partial_count_feat_3         -0.443038
perc_improv_failure_score_feat_3                 -0.502128
perc_improv_failure_score_normalized_feat_3      -0.405862
```
--------------
```
Groupwise Metrics: 
                                             count        mean       std  \
query_count                                    5.0  281.600000  0.894427   
old_ACR                                        5.0    1.650645  0.035191   
new_ACR                                        5.0    1.944711  0.120152   
old_MRR                                        5.0    0.785854  0.015501   
new_MRR                                        5.0    0.704742  0.018173   
old_failure_count_feat_3                       5.0    0.063916  0.016470   
new_failure_count_feat_3                       5.0    0.084509  0.008743   
old_failure_partial_count_feat_3               5.0    0.112221  0.021014   
new_failure_partial_count_feat_3               5.0    0.161958  0.013100   
old_failure_score_feat_3                       5.0    0.166895  0.018529   
new_failure_score_feat_3                       5.0    0.250790  0.035605   
old_failure_score_normalized_feat_3            5.0    0.088984  0.016843   
new_failure_score_normalized_feat_3            5.0    0.125114  0.006408   
perc_improv_ACR                                5.0  -17.787433  6.192517   
perc_improv_MRR                                5.0  -10.279926  3.420699   
perc_improv_failure_count_feat_3               5.0   -0.388731  0.366461   
perc_improv_failure_partial_count_feat_3       5.0   -0.472912  0.225234   
perc_improv_failure_score_feat_3               5.0   -0.514325  0.240581   
perc_improv_failure_score_normalized_feat_3    5.0   -0.443156  0.263466   

                                                    min         25%  \
query_count                                  281.000000  281.000000   
old_ACR                                        1.613475    1.614841   
new_ACR                                        1.794326    1.861210   
old_MRR                                        0.761744    0.782571   
new_MRR                                        0.682774    0.692215   
old_failure_count_feat_3                       0.046263    0.056738   
new_failure_count_feat_3                       0.071174    0.084806   
old_failure_partial_count_feat_3               0.088968    0.095745   
new_failure_partial_count_feat_3               0.148410    0.152482   
old_failure_score_feat_3                       0.142349    0.152482   
new_failure_score_feat_3                       0.208481    0.223404   
old_failure_score_normalized_feat_3            0.069649    0.079196   
new_failure_score_normalized_feat_3            0.116529    0.122462   
perc_improv_ACR                              -23.390558  -22.947368   
perc_improv_MRR                              -12.775388  -12.752412   
perc_improv_failure_count_feat_3              -0.846154   -0.687500   
perc_improv_failure_partial_count_feat_3      -0.800000   -0.592593   
perc_improv_failure_score_feat_3              -0.775000   -0.693878   
perc_improv_failure_score_normalized_feat_3   -0.808151   -0.597761   

                                                    50%         75%  \
query_count                                  281.000000  282.000000   
old_ACR                                        1.658363    1.676157   
new_ACR                                        1.943463    2.046263   
old_MRR                                        0.788061    0.793600   
new_MRR                                        0.702684    0.720858   
old_failure_count_feat_3                       0.056940    0.070671   
new_failure_count_feat_3                       0.085409    0.085409   
old_failure_partial_count_feat_3               0.116608    0.117438   
new_failure_partial_count_feat_3               0.160142    0.167260   
old_failure_score_feat_3                       0.174377    0.180212   
new_failure_score_feat_3                       0.252669    0.274021   
old_failure_score_normalized_feat_3            0.086833    0.095364   
new_failure_score_normalized_feat_3            0.125936    0.126537   
perc_improv_ACR                              -20.350109  -11.208791   
perc_improv_MRR                              -12.524845   -7.979601   
perc_improv_failure_count_feat_3              -0.250000   -0.200000   
perc_improv_failure_partial_count_feat_3      -0.424242   -0.275000   
perc_improv_failure_score_feat_3              -0.480769   -0.465116   
perc_improv_failure_score_normalized_feat_3   -0.410324   -0.221938   

                                                    max  
query_count                                  283.000000  
old_ACR                                        1.690391  
new_ACR                                        2.078292  
old_MRR                                        0.803295  
new_MRR                                        0.725177  
old_failure_count_feat_3                       0.088968  
new_failure_count_feat_3                       0.095745  
old_failure_partial_count_feat_3               0.142349  
new_failure_partial_count_feat_3               0.181495  
old_failure_score_feat_3                       0.185053  
new_failure_score_feat_3                       0.295374  
old_failure_score_normalized_feat_3            0.113879  
new_failure_score_normalized_feat_3            0.134104  
perc_improv_ACR                              -11.040340  
perc_improv_MRR                               -5.367386  
perc_improv_failure_count_feat_3               0.040000  
perc_improv_failure_partial_count_feat_3      -0.272727  
perc_improv_failure_score_feat_3              -0.156863  
perc_improv_failure_score_normalized_feat_3   -0.177604  
```
@asaintrequier

File: python/ml4ir/applications/ranking/model/metrics/metrics_helper.py [hunk 1]
```diff
@@ -3,16 +3,127 @@
 from typing import List
 
 
+def compute_failure_stats(
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/model/metrics/metrics_helper.py:6]: @asaintrequier This is the biggest code addition.

File: python/ml4ir/applications/ranking/model/metrics/metrics_helper.py [hunk 2]
```diff
@@ -3,16 +3,131 @@
...
+                    1 if (new_pre_click_secondary_label < click_secondary_label).all() else 0
+                )
+                # Query failure if failure on at least one record
+                new_failure_any = (
+                    1 if (new_pre_click_secondary_label < click_secondary_label).any() else 0
+                )
+                # Count of failure records
+                new_failure_count = (new_pre_click_secondary_label < click_secondary_label).sum()
+                # Normalizing to fraction of potential records
+                new_failure_fraction = new_failure_count / (new_click_rank - 1)
```
Comment (asaintrequier) [inline python/ml4ir/applications/ranking/model/metrics/metrics_helper.py:70]: You could factorise the computation of new & old metrics in a method

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/metrics/metrics_helper.py:70]: Yeah. I thought so too. Let me do it.

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/metrics/metrics_helper.py]: Done.

**Pull request merged**